src/lora.cpp

```diff
@@ -0,0 +1,152 @@
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ */
+#ifdef ENABLE_LORA
+
+#include <lmic.h>
+#include <hal/hal.h>
+#include <SPI.h>
+
+#include "emonesp.h"
+#include "lora.h"
+
+#include "app_config.h"
+
+
+#define LORA_HTOI(c) ((c<='9')?(c-'0'):((c<='F')?(c-'A'+10):((c<='f')?(c-'a'+10):(0))))
+#define LORA_TWO_HTOI(h, l) ((LORA_HTOI(h) << 4) + LORA_HTOI(l))
+#define LORA_HEX_TO_BYTE(a, h, n) { for (int i = 0; i < n; i++) (a)[i] = LORA_TWO_HTOI(h[2*i], h[2*i + 1]); }
+#define LORA_DEVADDR(a) (uint32_t) ((uint32_t) (a)[3] | (uint32_t) (a)[2] << 8 | (uint32_t) (a)[1] << 16 | (uint32_t) (a)[0] << 24)
+
+#define ANNOUNCE_INTERVAL 30 * 1000 // (In Milliseconds)
+
+
+// TODO: Store these via WebUI? We're doing (simple) ABP activation for now
+const char *devAddr = "00000000";
+const char *nwkSKey = "00000000000000000000000000000000";
+const char *appSKey = "00000000000000000000000000000000";
+
+
+// LoRaWAN credentials to use
+static uint8_t DEVADDR[4];
+static uint8_t NWKSKEY[16];
+static uint8_t APPSKEY[16];
+
+// LoRaWAN data packet
+unsigned long lastAnnounce;
+uint8_t dataPacket[9];
+
+// LoRa module pin mapping
+const lmic_pinmap lmic_pins = {
+  .nss = LORA_NSS,
+  .rxtx = LMIC_UNUSED_PIN,
+  .rst = LORA_RST,
+  .dio = {LORA_DIO0, LORA_DIO1, LORA_DIO2},
+};
+
+// Used for OTAA, not used (yet)
+void os_getArtEui (u1_t* buf) { }
+void os_getDevEui (u1_t* buf) { }
+void os_getDevKey (u1_t* buf) { }
+
+
+void onEvent(ev_t ev) {
+  switch (ev) {
+    case EV_TXCOMPLETE:
+      DBUGF("LoRa: TX Complete.");
+      // LoRaWAN transmission complete
+      if (LMIC.txrxFlags & TXRX_ACK) {
+        // Received ack
+        DBUGF("LoRa: TX ack.");
+      }
+      break;
+    case EV_TXSTART:
+      DBUGF("LoRa: TX Begin.");
+      break;
+  }
+}
+
+/// Reset LoRa modem. Reload LoRaWAN keys
+void lora_reset()
+{
+  LORA_HEX_TO_BYTE(DEVADDR, devAddr, 4);
+  LORA_HEX_TO_BYTE(NWKSKEY, nwkSKey, 16);
+  LORA_HEX_TO_BYTE(APPSKEY, appSKey, 16);
+
+  LMIC_reset();
+  LMIC_setSession (0x13, LORA_DEVADDR(DEVADDR), NWKSKEY, APPSKEY);
+  LMIC_setAdrMode(0);
+  LMIC_setClockError(MAX_CLOCK_ERROR * 10 / 100);
+  LMIC_selectSubBand(1);
+  LMIC_setLinkCheckMode(0);
+  LMIC.dn2Dr = DR_SF7;
+}
+
+
+/// Initial setup of LoRa modem.
+void lora_setup()
+{
+  Profile_Start(lora_setup);
+
+  os_init();
+  lora_reset();
+
+  // Set us up for an immeadiate announcement
+  lastAnnounce = millis() - ANNOUNCE_INTERVAL;
+ 
+  Profile_End(lora_setup, 1);
+}
+
+
+void lora_loop()
+{
+  Profile_Start(lora_loop);
+
+  if (lastAnnounce < lastAnnounce + ANNOUNCE_INTERVAL)
+    return;
+
+  DBUGF("LoRa: Starting LoRaWAN broadcast...");
+  // Check if there is not a current TX/RX job running
+  if (LMIC.opmode & OP_TXRXPEND) {
+    DBUGF("LoRa: Modem busy. Retry later");
+    return;
+  } 
+
+  // TODO: Pack important OpenEVSE stats into dataPacket.
+
+  LMIC_setTxData2(1, dataPacket, sizeof(dataPacket), true);
+
+  lastAnnounce = millis();
```

Comment:
IHMO it is better to work out what time the next packet should be, ie:
```suggestion
  nextAnnounce = millis() + ANNOUNCE_INTERVAL;
```

You only need to do the addition once rather than every check, but a minor point

Reply:
Ah. Good point :-)